src/Data/String/CodePoints.purs

```diff
@@ -0,0 +1,273 @@
+
+
+-- | Returns the number of code points in the given string. Operates in
+-- | constant space and time linear to the length of the string.
+length :: String -> Int
+length = Array.length <<< toCodePointArray
+
+
+-- | Creates a string containing just the given code point. Operates in
+-- | constant space and time.
+singleton :: CodePoint -> String
+singleton = _singleton singletonFallback
+
+foreign import _singleton
+  :: (CodePoint -> String)
+  -> CodePoint
+  -> String
+
+singletonFallback :: CodePoint -> String
+singletonFallback (CodePoint cp) | cp <= 0xFFFF = fromCharCode cp
+singletonFallback (CodePoint cp) =
+  let lead = ((cp - 0x10000) / 0x400) + 0xD800 in
+  let trail = (cp - 0x10000) `mod` 0x400 + 0xDC00 in
+  fromCharCode lead <> fromCharCode trail
+
+
+-- | Returns a record with strings created from the code points on either side
+-- | of the given index. If the index is not within the string, Nothing is
+-- | returned.
+splitAt :: Int -> String -> Maybe { before :: String, after :: String }
+splitAt i s =
+  let cps = toCodePointArray s in
+  if i < 0 || Array.length cps < i
+    then Nothing
+    else Just {
+        before: fromCodePointArray (Array.take i cps),
+        after: fromCodePointArray (Array.drop i cps)
+      }
+
+
+-- | Returns a string containing the given number of code points from the
+-- | beginning of the given string. If the string does not have that many code
+-- | points, returns the empty string. Operates in space and time linear to the
+-- | given number.
+take :: Int -> String -> String
+take = _take takeFallback
+
+foreign import _take :: (Int -> String -> String) -> Int -> String -> String
+
+takeFallback :: Int -> String -> String
+takeFallback n _ | n < 1 = ""
+takeFallback n s = case uncons s of
+  Just { head, tail } -> singleton head <> takeFallback (n - 1) tail
+  _ -> s
+
+
+-- | Returns a string containing the leading sequence of code points which all
+-- | match the given predicate from the given string. Operates in space and
+-- | time linear to the given number.
+takeWhile :: (CodePoint -> Boolean) -> String -> String
+takeWhile p s = take (count p s) s
+
+
+-- | Creates an array of code points from a string. Operates in space and time
+-- | linear to the length of the given string.
+toCodePointArray :: String -> Array CodePoint
+toCodePointArray = _toCodePointArray toCodePointArrayFallback unsafeCodePointAt0
+
+foreign import _toCodePointArray
+  :: (String -> Array CodePoint)
+  -> (String -> CodePoint)
+  -> String
+  -> Array CodePoint
+
+toCodePointArrayFallback :: String -> Array CodePoint
+toCodePointArrayFallback s = unfoldr decode s
+  where
+  decode :: String -> Maybe (Tuple CodePoint String)
+  decode s' = (\{ head, tail } -> Tuple head tail) <$> uncons s'
+
+
+-- | Returns a record with the first code point and the remaining code points
+-- | of the given string. Returns Nothing if the string is empty. Operates in
+-- | space and time linear to the length of the string.
```

Comment:
I was operating under the assumption that a naïve implementation of `String.drop` would copy the remainder of the string. But that's actually probably not the case. I can do tests, but I'm comfortable just asserting it will be constant.